cmd/soroban-rpc/internal/db/transaction.go

```diff
@@ -0,0 +1,336 @@
+		return errors.New("TransactionWriter incorrectly initialized without stmtCache")
+	} else if txCount == 0 {
+		return nil
+	}
+
+	reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(txn.passphrase, lcm)
+	if err != nil {
+		return errors.Wrapf(err,
+			"failed to open transaction reader for ledger %d",
+			lcm.LedgerSequence())
+	}
+
+	transactions := make(map[xdr.Hash]ingest.LedgerTransaction, txCount)
+	for i := 0; i < txCount; i++ {
+		tx, err := reader.Read()
+		if err != nil {
+			return errors.Wrapf(err, "failed reading tx %d", i)
+		}
+
+		// For fee-bump transactions, we store lookup entries for both the outer
+		// and inner hashes.
+		if tx.Envelope.IsFeeBump() {
+			transactions[tx.Result.InnerHash()] = tx
+		}
+		transactions[tx.Result.TransactionHash] = tx
+	}
+
+	mid = time.Now()
+	query := sq.Insert(transactionTableName).
+		Columns("hash", "ledger_sequence", "application_order")
+	for hash, tx := range transactions {
+		query = query.Values(hash[:], lcm.LedgerSequence(), tx.Index)
+	}
+	_, err = query.RunWith(txn.stmtCache).Exec()
+
+	L.WithError(err).
+		WithField("duration", time.Since(start)).
+		WithField("sql_duration", time.Since(mid)).
+		Infof("Ingested %d transaction lookups", len(transactions))
+
+	return err
+}
+
+func (txn *transactionHandler) RegisterMetrics(ingest, insert, count prometheus.Observer) {
+	txn.ingestMetric = ingest
+	txn.insertMetric = insert
+	txn.countMetric = count
+}
+
+// trimTransactions removes all transactions which fall outside the ledger retention window.
+func (txn *transactionHandler) trimTransactions(latestLedgerSeq uint32, retentionWindow uint32) error {
+	if latestLedgerSeq+1 <= retentionWindow {
+		return nil
+	}
+
+	cutoff := latestLedgerSeq + 1 - retentionWindow
+	_, err := sq.StatementBuilder.
+		RunWith(txn.stmtCache).
+		Delete(transactionTableName).
+		Where(sq.Lt{"ledger_sequence": cutoff}).
+		Exec()
+	return err
+}
+
+// GetLedgerRange pulls the min/max ledger sequence numbers from the database.
+func (txn *transactionHandler) GetLedgerRange(ctx context.Context) (ledgerbucketwindow.LedgerRange, error) {
+	var ledgerRange ledgerbucketwindow.LedgerRange
+
+	//
+	// We use subqueries alongside a UNION ALL stitch in order to select the min
+	// and max from the ledger table in a single query and get around sqlite's
+	// limitations with parentheses (see https://stackoverflow.com/a/22609948).
+	//
+	newestQ := sq.
+		Select("m1.meta").
+		FromSelect(
+			sq.
+				Select("meta").
+				From(ledgerCloseMetaTableName).
+				OrderBy("sequence ASC").
+				Limit(1),
+			"m1",
+		)
+	oldestQ := sq.
+		Select("m2.meta").
+		FromSelect(
+			sq.
+				Select("meta").
+				From(ledgerCloseMetaTableName).
+				OrderBy("sequence DESC").
+				Limit(1),
+			"m2",
+		)
+
+	sql, args, err := oldestQ.ToSql()
+	if err != nil {
+		return ledgerRange, errors.Wrap(err, "couldn't build ledger range query")
+	}
+
+	rows, err := txn.db.Query(ctx, newestQ.Suffix(" UNION ALL "+sql, args...))
```

Comment:
you can use the `txn.db.Select()` function to get all the rows and that would allow you to simplify the error checking  because you wouldn't need to call `rows.Next() ` and `rows.Scan()` twice

Reply:
Thanks for the tip! I figured it might've been possible but since there weren't any docs I had trouble finding the right syntax combination - gave it another go and fixed it up in 57f6f78.